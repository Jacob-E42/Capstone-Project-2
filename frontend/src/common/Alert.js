import React, { useCallback, useContext, useState } from "react";
import { Alert as MuiAlert, Snackbar } from "@mui/material";
import AlertContext from "../context_providers/AlertContext";

function Alert({ msg, color = "success" }) {
	const { setMsg } = useContext(AlertContext);
	console.debug("Alert", "messages=", msg, "color=", color);

	// State to control the visibility of the alert message
	const [visible, setVisible] = useState(true);

	// Function to dismiss the alert message
	const onDismiss = useCallback(() => {
		setVisible(false);
		setMsg("");
	}, [setVisible, setMsg]);

	// return (
	// 	<Alert
	// 		// Set the color of the alert message
	// 		color={color}
	// 		// Specify if the alert is visible or hidden
	// 		isOpen={visible}
	// 		// Function to toggle the visibility of the alert
	// 		toggle={onDismiss}>
	// 		{msg}
	// 	</Alert>
	// );
	return (
		<Snackbar
			open={visible}
			autoHideDuration={6000}
			onClose={onDismiss}>
			<MuiAlert
				onClose={onDismiss}
				severity={color}
				sx={{ width: "100%" }}>
				{msg}
			</MuiAlert>
		</Snackbar>
	);
}

export default Alert;
