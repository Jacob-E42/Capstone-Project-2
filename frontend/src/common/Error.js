import React from "react";
import { Alert } from "reactstrap";

const Error = ({ msg, color }) => {
	return <Alert color={color}>{msg}</Alert>;
};

export default Error;
