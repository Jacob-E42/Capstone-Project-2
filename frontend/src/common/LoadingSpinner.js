import React from "react";
import { CircularProgress, Box } from "@mui/material";

function LoadingSpinner() {
	return (
		<Box sx={{ display: "flex", justifyContent: "center", p: 3 }}>
			<CircularProgress />
		</Box>
	);
}

export default LoadingSpinner;
