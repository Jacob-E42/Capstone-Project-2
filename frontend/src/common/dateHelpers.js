import {
	eachDayOfInterval,
	format,
	parseISO,
	subDays,
	getDay,
	differenceInCalendarDays,
	addDays,
	isToday
} from "date-fns";

// Return today's date in format "yyyy-mm-dd"
function getCurrentDate() {
	const currentDate = format(new Date(), "yyyy-MM-dd", { weekStartsOn: 0 });
	return currentDate;
}

//Return an array of Dates corresponding to the difference between the calendar day of dateString and the date
// that's num days before the dateString.
// The date passed in is included as the last element in the array. Therefore the number of returned elements is n + 1.
//
//dateString has format: "yyyy-mm-dd"
const getPreviousNumDays = (dateString, num) => {
	const date = parseISO(dateString);
	// console.log("date=", dateString, date);
	let days = eachDayOfInterval({ start: subDays(date, num), end: date });
	const daysStrings = days.map(date => {
		return format(date, "yyyy-MM-dd");
	});
	return daysStrings;
};
//Return an array of Dates corresponding to the difference between the calendar day of dateString and the date
// that's num days after the dateString.
//dateString has format: "yyyy-mm-dd"
const getNextNumDays = (dateString, num) => {
	if (num === 0) return [];
	const date = parseISO(dateString);
	let days = eachDayOfInterval({ start: addDays(date, 1), end: addDays(date, num) });
	const daysStrings = days.map(date => {
		return format(date, "yyyy-MM-dd");
	});

	return daysStrings;
};

//Return a string corresponding to the day of the week of the dateString
//dateString has format: "yyyy-mm-dd"
const getDayOfWeek = dateString => {
	const days = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];
	const date = parseISO(dateString);

	// const day = days[getDay(date)];
	// console.log(day, getDay(date));
	return days[getDay(date)];
};

//get the range of dateStrings that the StreakSlider should render
// accepts the dateString of the day whose journal will be shown
function getDateRange(dateString) {
	const date = parseISO(dateString);
	const currentDate = parseISO(getCurrentDate());

	const difference = differenceInCalendarDays(currentDate, date);
	if (difference < 0) return new RangeError("The date cannot be in the future!");
	const prevDays = getPreviousNumDays(dateString, 30);
	const nextDays = difference <= 30 ? getNextNumDays(dateString, difference) : getNextNumDays(dateString, 30);

	const dateRange = [...prevDays, ...nextDays];

	// console.log("difference=", difference, "prevDays=", prevDays, "nextDays=", nextDays);
	return dateRange;
}

function getPastDate(dateString, num) {
	const currentDate = parseISO(dateString);
	let pastDate = subDays(currentDate, num);
	return format(pastDate, "yyyy-MM-dd");
}

function getFutureDate(dateString, num) {
	const currentDate = parseISO(dateString);
	let fututeDate = addDays(currentDate, num);
	return format(fututeDate, "yyyy-MM-dd");
}

function isCurrentDate(dateString) {
	const currentDate = parseISO(dateString);

	return isToday(currentDate);
}

export {
	getPreviousNumDays,
	getDayOfWeek,
	getCurrentDate,
	getNextNumDays,
	getDateRange,
	getPastDate,
	getFutureDate,
	isCurrentDate
};
