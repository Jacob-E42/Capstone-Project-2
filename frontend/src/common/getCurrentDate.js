function getCurrentDate() {
	const currentDate = new Date().toISOString().slice(0, 10); // e.g., "2023-07-25"
	return currentDate;
}

export default getCurrentDate;
