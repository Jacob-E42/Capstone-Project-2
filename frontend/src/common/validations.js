function validateDateUserIdAndApi(date, userId, api) {
	if (!date || !userId || !api) return false;
	if (typeof date !== "string") return false;
	if (date.length < 10) return false;
	if (typeof userId !== "number") return false;
	if (typeof api !== "object") return false;
	return true;
}

function validateDate(date) {
	if (!date) return false;
	if (typeof date !== "string") return false;
	if (date.length < 10) return false;

	return true;
}

const validateDateUserAndApi = (date, user, api) => {
	if (!(date && user && api)) return false;
	if (typeof date !== "string") return false;
	if (date.length < 10) return false;
	if (typeof user !== "object") return false;
	if (typeof api !== "object") return false;
	return true;
};

const validateJournalInfo = (id, userId, title, entryText, date, journalType) => {
	console.debug("validateJournalInfo", id, userId, title, entryText, date, journalType);
	if (!id || !userId || !title || !entryText || !date || !journalType)
		return { valid: false, error: "REQUIRED INFO IS MISSING" };
	if (typeof id !== "number") return { valid: false, error: "ID IS NOT A NUMBER" };
	if (typeof userId !== "number") return { valid: false, error: "USERID IS NOT A NUMBER" };
	if (typeof title !== "string") return { valid: false, error: "TITLE IS NOT A STRING" };
	if (typeof date !== "string") return { valid: false, error: "DATE IS NOT A STRING" };
	if (typeof entryText !== "string") return { valid: false, error: "ENTRYTEXT IS NOT A STRING" };
	if (typeof journalType !== "string") return { valid: false, error: "JOURNALTYPE IS NOT A STRING" };
	if (entryText.trim().length < 1) return { valid: false, error: "ENTRYTEXT IS TOO SHORT" };
	if (date.length !== 10) return { valid: false, error: "DATE IS WRONG LENGTH" };
	if (!journalType.includes("Journal")) return { valid: false, error: "JOURNAL TYPE IS WRONG" };

	return { valid: true };
};

export { validateDateUserAndApi, validateDateUserIdAndApi, validateJournalInfo, validateDate };
