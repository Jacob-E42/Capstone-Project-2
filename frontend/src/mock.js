import React, { useCallback, useState } from "react";
import UserContext from "./context_providers/UserContext";
import AlertContext from "./context_providers/AlertContext";
import Request from "./api";
import ApiRequest from "./api";
import ApiContext from "./context_providers/ApiContext";
import JournalContext from "./context_providers/JournalContext";

const demoUser = {
	id: 11,
	firstName: "testfirst",
	lastName: "testlast",
	email: "test@test.net",
	interests: ["baking", "jetskiing"]
};

const demoJournal = {
	id: 42,
	userId: 11,
	title: "Baking adventure",
	entryText:
		"Yesterday I tried baking for the first time like I've always wanted to. It was a complete disaster, but at least it was fun.",
	entryDate: "2023-07-24",
	emotions: null
};

const UserProvider = ({ children }) => {
	return <UserContext.Provider value={{ user: demoUser }}>{children}</UserContext.Provider>;
};

const AnonUserProvider = ({ children }) => {
	const [currentUser, setCurrentUser] = useState(null);

	const login = useCallback(
		async formData => {
			try {
				const request = new Request();
				const token = await request.login(formData);
				if (token) {
					setCurrentUser(demoUser);
					return { success: true };
				} else throw new Error("Log in token in missing");
			} catch (errors) {
				console.error("login failed", errors);
				return { success: false, errors };
			}
		},
		[setCurrentUser]
	);

	const signup = useCallback(
		async formData => {
			try {
				const request = new Request();
				const token = await request.signup(formData);
				console.log(token);
				if (token) setCurrentUser(demoUser);
				else throw new Error("missing sign in token");
				return { success: true };
			} catch (errors) {
				console.error("signup failed", errors);
				return { success: false, errors };
			}
		},
		[setCurrentUser]
	);

	return <UserContext.Provider value={{ currentUser, login, signup }}>{children}</UserContext.Provider>;
};

const AlertProvider = ({ children }) => {
	const [msg, setMsg] = useState("");
	const [color, setColor] = useState("primary");

	return <AlertContext.Provider value={{ msg, setMsg, color, setColor }}>{children}</AlertContext.Provider>;
};

const ApiProvider = ({ children }) => {
	const [request, setRequest] = useState(new ApiRequest());
	request.setToken(
		`eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MTEsImVtYWlsIjoidGVzdEB0ZXN0Lm5ldCIsImlhdCI6MTY5MjI5MDkxN30.8e12Cw-TMts0FmZALH61ch03kUMlW5QoCi60dlfhyWY`
	);

	return <ApiContext.Provider value={{ api: request }}>{children}</ApiContext.Provider>;
};

const JournalProvider = ({ children }) => {
	return <JournalContext.Provider value={{ journal: demoJournal }}>{children}</JournalContext.Provider>;
};

export { UserProvider, AnonUserProvider, AlertProvider, ApiProvider, JournalProvider };
