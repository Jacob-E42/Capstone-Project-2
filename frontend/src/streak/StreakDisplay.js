import React, { useEffect } from "react";
// import Slider from "react-slick";
import "slick-carousel/slick/slick.css";
import "slick-carousel/slick/slick-theme.css";
import "./Streak.css";
// import DayBlock from "./DayBlock";
import useLocalStorage from "../hooks/useLocalStorage";
import useValidateDate from "../hooks/useVerifyDate";
import { getCurrentDate, getDateRange, getDayOfWeek } from "../common/dateHelpers";
import { format, parseISO } from "date-fns";
import { useNavigate, Link } from "react-router-dom";
import LoadingSpinner from "../common/LoadingSpinner";

const StreakDisplay = ({ date }) => {
	// console.debug("StreakDisplay", date);
	const navigate = useNavigate();
	if (!useValidateDate(date)) navigate(`journal/${getCurrentDate()}`);
	const [daysArray, setDaysArray] = useLocalStorage("datesToRender", [date]);
	const [slidesLoaded, setSlidesLoaded] = useLocalStorage("slidesLoaded", false);
	// const startDate = daysArray[0];
	// const endDate = daysArray[daysArray.length - 1];
	// const daysArray = eachDayOfInterval({ start: new Date(startDate), end: new Date(endDate) }).reverse(); // reverse to have endDate at one end
	useEffect(() => {
		setDaysArray(getDateRange(date));
		setSlidesLoaded(true);
	}, [date, setDaysArray, setSlidesLoaded]);

	// const daySlides = datesToRender.map((day, index) => {
	// 	return (
	// 		<DayBlock
	// 			day={day}
	// 			key={index}
	// 		/>
	// 	);
	// });
	// console.log(datesToRender, datesToRender.length, typeof daySlides);

	if (!slidesLoaded) return <LoadingSpinner />;

	// const totalSlides = datesToRender.length;

	// const settings = {
	// 	accessability: true,
	// 	arrows: true,
	// 	centerMode: false,
	// 	className: "Slider",
	// 	dots: true,
	// 	infinite: false,
	// 	initialSlide: datesToRender.length - 1,
	// 	slide: "div",
	// 	speed: 500,
	// 	slidesToShow: process.env.NODE_ENV === "test" ? totalSlides : 10,
	// 	slidesToScroll: 3
	// };

	return (
		<div className="streak-container">
			{daysArray.map(day => (
				<Link
					key={day}
					to={`/journal/${day}`}
					className={`day-link ${day === date ? "active" : ""}`}>
					{getDayOfWeek(day)}
					<br />
					{format(parseISO(day), "dd")}
				</Link>
			))}
		</div>
	);

	// return (
	// 	<div className="StreakDisplay">

	// 		<Slider {...settings}>{daySlides}</Slider>
	// 	</div>
	// );
};

export default StreakDisplay;
