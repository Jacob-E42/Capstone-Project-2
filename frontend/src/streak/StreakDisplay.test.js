import React from "react";
import { screen, render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { AlertProvider, UserProvider } from "../mock";
import StreakDisplay from "./StreakDisplay";
import { getCurrentDate, getPastDate } from "../common/dateHelpers";
window.HTMLElement.prototype.scrollIntoView = function () {};

test("StreakDisplay renders without crashing", () => {
	render(
		<MemoryRouter>
			<UserProvider>
				<AlertProvider>
					<StreakDisplay date={getCurrentDate()} />
				</AlertProvider>
			</UserProvider>
		</MemoryRouter>
	);
});

test("StreakDisplay matches snapshot", () => {
	const { asFragment } = render(
		<MemoryRouter>
			<UserProvider>
				<AlertProvider>
					<StreakDisplay date={getCurrentDate()} />
				</AlertProvider>
			</UserProvider>
		</MemoryRouter>
	);
	expect(asFragment()).toMatchSnapshot();
});

test("StreakDisplay renders 30 days when date is today", () => {
	render(
		<MemoryRouter>
			<UserProvider>
				<AlertProvider>
					<StreakDisplay date={getCurrentDate()} />
				</AlertProvider>
			</UserProvider>
		</MemoryRouter>
	);

	const allDaysRendered = screen.getAllByRole("link");
	// console.log(allDaysRendered);
	expect(allDaysRendered.length).toBe(31);
});

test("StreakDisplay renders at least 60 days when date is over a month ago", () => {
	const today = getCurrentDate();
	const pastDate = getPastDate(today, 45);
	render(
		<MemoryRouter>
			<UserProvider>
				<AlertProvider>
					<StreakDisplay date={pastDate} />
				</AlertProvider>
			</UserProvider>
		</MemoryRouter>
	);

	const allDaysRendered = screen.getAllByRole("link");
	console.log(allDaysRendered.length);
	expect(allDaysRendered.length).toBeGreaterThan(60);
});

test("StreakDisplay renders correct amount of days when date is less than a month ago", () => {
	const today = getCurrentDate();
	const pastDate = getPastDate(today, 15);
	render(
		<MemoryRouter>
			<UserProvider>
				<AlertProvider>
					<StreakDisplay date={pastDate} />
				</AlertProvider>
			</UserProvider>
		</MemoryRouter>
	);

	const allDaysRendered = screen.getAllByRole("link");
	console.log(allDaysRendered.length);
	expect(allDaysRendered.length).toBe(46);
});
