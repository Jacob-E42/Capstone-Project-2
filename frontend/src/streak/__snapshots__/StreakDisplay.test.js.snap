// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StreakDisplay matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="streak-container"
  >
    <a
      class="day-link "
      href="/journal/2023-10-01"
    >
      Sun
      <br />
      01
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-02"
    >
      Mon
      <br />
      02
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-03"
    >
      Tue
      <br />
      03
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-04"
    >
      Wed
      <br />
      04
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-05"
    >
      Thu
      <br />
      05
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-06"
    >
      Fri
      <br />
      06
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-07"
    >
      Sat
      <br />
      07
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-08"
    >
      Sun
      <br />
      08
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-09"
    >
      Mon
      <br />
      09
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-10"
    >
      Tue
      <br />
      10
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-11"
    >
      Wed
      <br />
      11
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-12"
    >
      Thu
      <br />
      12
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-13"
    >
      Fri
      <br />
      13
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-14"
    >
      Sat
      <br />
      14
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-15"
    >
      Sun
      <br />
      15
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-16"
    >
      Mon
      <br />
      16
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-17"
    >
      Tue
      <br />
      17
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-18"
    >
      Wed
      <br />
      18
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-19"
    >
      Thu
      <br />
      19
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-20"
    >
      Fri
      <br />
      20
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-21"
    >
      Sat
      <br />
      21
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-22"
    >
      Sun
      <br />
      22
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-23"
    >
      Mon
      <br />
      23
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-24"
    >
      Tue
      <br />
      24
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-25"
    >
      Wed
      <br />
      25
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-26"
    >
      Thu
      <br />
      26
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-27"
    >
      Fri
      <br />
      27
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-28"
    >
      Sat
      <br />
      28
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-29"
    >
      Sun
      <br />
      29
    </a>
    <a
      class="day-link "
      href="/journal/2023-10-30"
    >
      Mon
      <br />
      30
    </a>
    <a
      class="day-link active"
      href="/journal/2023-10-31"
    >
      Tue
      <br />
      31
    </a>
  </div>
</DocumentFragment>
`;
